### PR TITLE
New segmentation loss functions and metrics

### PR DESCRIPTION
**Pull request by jlcsilva**
Signed-off-by: João Lourenço Silva <joao.lourenco.silva@tecnico.ulisboa.pt>

### Description
Added Generalized Dice Focal Loss and Generalized Dice Score metric. Implemented function in metrics.utils to check whether tensors are binarized and refactored existing metrics to use it.

### Status
**Ready**

### Types of changes

- [] Non-breaking change (fix or new feature that would not break existing functionality).
- [x ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ x] New tests added to cover the changes.
- [ x] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ x] In-line docstrings updated.
- [ x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/metrics/confusion_matrix.py [hunk 1]
```diff
@@ -84,13 +84,15 @@ def _compute_tensor(self, y_pred: torch.Tensor, y: torch.Tensor):  # type: ignor
...
-            raise ValueError("y_pred and y must be PyTorch Tensor.")
-        # check binarized input
-        if not torch.all(y_pred.byte() == y_pred):
-            warnings.warn("y_pred should be a binarized tensor.")
-        if not torch.all(y.byte() == y):
-            raise ValueError("y should be a binarized tensor.")
+        try:
+            is_binary_tensor(y_pred, "y_pred")
+        except ValueError as e:
+            raise ValueError("y_pred" + str(e)) from e
```
Comment (wyli) [inline monai/metrics/confusion_matrix.py:90]: cool, I think we can drop the try...except here:
```suggestion
        is_binary_tensor(y_pred, "y_pred")
```

Comment (jlcsilva) [reply monai/metrics/confusion_matrix.py]: Yeah we can, my bad 😅

**Review by wyli — APPROVED** [review]
Thank you, this improves metrics and  losses and with all changes verified.

**Pull request merged**